File: native-plugins/admob/resources/android/src/ConsentManager.kt
package com.blessedzulu.nativephp.admob

import android.content.Context
import androidx.fragment.app.FragmentActivity
import com.google.android.ump.ConsentDebugSettings
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.UserMessagingPlatform

/**
 * Holds the process-wide UMP ConsentInformation singleton and the helpers that
 * surround it: debug-settings construction (so the consent form can be forced
 * on a non-EEA test device) and the int-status -> PHP-string mapping that keeps
 * the native layer aligned with ConsentChanged::STATUS_* on the Laravel side.
 *
 * Mirrors the registry pattern used by the ad formats - a single object that
 * owns one piece of SDK state for the whole app.
 */
object ConsentManager {
    @Volatile
    private var cached: ConsentInformation? = null

    @Synchronized
    fun info(context: Context): ConsentInformation =
        cached ?: UserMessagingPlatform.getConsentInformation(context.applicationContext)
            .also { cached = it }

    fun requestParameters(context: Context, debugGeography: String): ConsentRequestParameters {
        val builder = ConsentRequestParameters.Builder()
            .setTagForUnderAgeOfConsent(AdmobInit.underAgeOfConsent())
        buildDebugSettings(context, debugGeography)?.let { builder.setConsentDebugSettings(it) }

        return builder.build()
    }

    /**
     * Builds ConsentDebugSettings from env so the consent form's geography can
     * be forced during testing. ADMOB_UMP_DEBUG_GEOGRAPHY is one of
     * EEA / NOT_EEA / DISABLED. (Test devices are managed in the AdMob console,
     * not here; the UMP debug geography honours a registered console test device
     * or simply use a VPN to a EEA region to exercise the real form.)
     */
    private fun buildDebugSettings(context: Context, debugGeography: String): ConsentDebugSettings? {
        val geography = debugGeography.trim().uppercase()

        if (geography.isBlank() || geography == "DISABLED") {
            return null
        }

        val builder = ConsentDebugSettings.Builder(context)
        when (geography) {
            "EEA" -> builder.setDebugGeography(ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_EEA)
            "NOT_EEA", "NON_EEA" -> builder.setDebugGeography(ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_NOT_EEA)
            else -> { /* leave default geography */ }
        }

        return builder.build()
    }

    /**
     * Maps ConsentInformation.consentStatus (int) to the PHP string constants in
     * BlessedZulu\NativePhpAdmob\Events\ConsentChanged.
     */
    fun statusString(info: ConsentInformation): String = when (info.consentStatus) {
        ConsentInformation.ConsentStatus.NOT_REQUIRED -> "not_required"
        ConsentInformation.ConsentStatus.REQUIRED -> "required"
        ConsentInformation.ConsentStatus.OBTAINED -> "obtained"
        else -> "unknown"
    }

    fun isFormRequired(info: ConsentInformation): Boolean =
        info.consentStatus == ConsentInformation.ConsentStatus.REQUIRED

    fun privacyOptionsRequired(info: ConsentInformation): Boolean =
        info.privacyOptionsRequirementStatus == ConsentInformation.PrivacyOptionsRequirementStatus.REQUIRED

    fun privacyOptionsStatusString(info: ConsentInformation): String =
        when (info.privacyOptionsRequirementStatus) {
            ConsentInformation.PrivacyOptionsRequirementStatus.REQUIRED -> "required"
            ConsentInformation.PrivacyOptionsRequirementStatus.NOT_REQUIRED -> "not_required"
            else -> "unknown"
        }
}

File: native-plugins/admob/resources/ios/Sources/ConsentManager.swift
import Foundation
import UserMessagingPlatform

/**
 * Surrounds the UMP ConsentInformation singleton: debug-settings construction
 * (so the consent form can be forced during testing) and the consentStatus ->
 * PHP-string mapping aligned with ConsentChanged::STATUS_* on the Laravel side.
 *
 * NOTE: GoogleUserMessagingPlatform ~> 2.7 exposes the UMP-prefixed ObjC type
 * names to Swift (UMPConsentInformation / UMPConsentForm / UMPRequestParameters
 * / UMPDebugSettings / UMPDebugGeography / UMPConsentStatus), and the singleton
 * is `.sharedInstance` (not `.shared`). Verified against a real compile
 * (Xcode 26, iOS Simulator) 2026-07-06. The enum cases (.EEA / .notEEA /
 * .disabled, .notRequired / .required / .obtained) import de-prefixed and are
 * resolved from their property/parameter type, so they need no UMP prefix.
 */
enum ConsentManager {
    static var info: UMPConsentInformation { UMPConsentInformation.sharedInstance }

    static func requestParameters(debugGeography: String) -> UMPRequestParameters {
        let params = UMPRequestParameters()
        params.tagForUnderAgeOfConsent = AdmobInit.underAgeOfConsent
        if let debug = debugSettings(debugGeography: debugGeography) {
            params.debugSettings = debug
        }

        return params
    }

    /**
     * Builds DebugSettings from env so the consent form's geography can be
     * forced during testing. ADMOB_UMP_DEBUG_GEOGRAPHY is one of
     * EEA / NOT_EEA / DISABLED. (Test devices are managed in the AdMob console,
     * not here; or simply use a VPN to a EEA region to exercise the real form.)
     */
    private static func debugSettings(debugGeography: String) -> UMPDebugSettings? {
        let geography = debugGeography.trimmingCharacters(in: .whitespaces).uppercased()

        if geography == "" || geography == "DISABLED" {
            return nil
        }

        let settings = UMPDebugSettings()
        switch geography {
        case "EEA": settings.geography = .EEA
        case "NOT_EEA", "NON_EEA": settings.geography = .notEEA
        default: settings.geography = .disabled
        }

        return settings
    }

    /**
     * Maps ConsentInformation.consentStatus to the PHP string constants in
     * BlessedZulu\NativePhpAdmob\Events\ConsentChanged.
     */
    static func statusString() -> String {
        switch info.consentStatus {
        case .notRequired: return "not_required"
        case .required: return "required"
        case .obtained: return "obtained"
        default: return "unknown"
        }
    }

    static func isFormRequired() -> Bool {
        return info.consentStatus == .required
    }

    static func privacyOptionsRequired() -> Bool {
        return info.privacyOptionsRequirementStatus == .required
    }

    static func privacyOptionsStatusString() -> String {
        switch info.privacyOptionsRequirementStatus {
        case .required: return "required"
        case .notRequired: return "not_required"
        default: return "unknown"
        }
    }
}
